controlplane/internal/controller/agentclusterinstall_controller.go

```diff
@@ -58,78 +53,70 @@ func (r *AgentClusterInstallReconciler) Reconcile(ctx context.Context, req ctrl.
+		}
+	}
+	return ctrl.Result{}, nil
+}
 
-			log.Info("Agent cluster install for control plane nodes has finished. Attaching kubeconfig secret to agent control plane",
-				"agent cluster install name", aci.Name, "agent cluster install namespace", aci.Namespace, "agent control plane name", acp.Name)
+func (r *AgentClusterInstallReconciler) reconcileInitialized(ctx context.Context, aci *hiveext.AgentClusterInstall, acp *controlplanev1alpha1.AgentControlPlane) error {
+	if !hasKubeconfigRef(aci) {
+		return nil
+	}
 
-			kubeconfigSecret, err := r.getACIKubeconfig(ctx, aci, acp)
-			if err != nil {
-				log.Error(err, "failed getting kubeconfig secret for agentclusterinstall", "agent cluster install name", aci.Name, "agent cluster install namespace", aci.Namespace, "secret name", kubeconfigSecret.Name)
-				return ctrl.Result{}, err
-			}
-			log.Info("found kubeconfig secret", "agent cluster install name", aci.Name, "agent cluster install namespace", aci.Namespace, "secret name", kubeconfigSecret.Name)
+	kubeconfigSecret, err := r.getACIKubeconfig(ctx, aci, *acp)
+	if err != nil {
+		return err
+	}
 
-			clusterName := acp.Labels[clusterv1.ClusterNameLabel]
-			labels := map[string]string{
-				clusterv1.ClusterNameLabel: clusterName,
-			}
+	clusterName := acp.Labels[clusterv1.ClusterNameLabel]
+	labels := map[string]string{
+		clusterv1.ClusterNameLabel: clusterName,
+	}
 
-			if err := r.updateLabels(ctx, kubeconfigSecret, labels); err != nil {
-				log.Error(err, "failed updating kubeconfig secret for agentclusterinstall", "agent cluster install name", aci.Name, "agent cluster install namespace", aci.Namespace, "secret name", kubeconfigSecret.Name)
-				return ctrl.Result{}, err
-			}
+	if err := r.updateLabels(ctx, kubeconfigSecret, labels); err != nil {
+		return err
+	}
 
-			if !r.ClusterKubeconfigSecretExists(ctx, clusterName, acp.Namespace) {
-				log.Info("Cluster kubeconfig doesn't exist, creating it")
-				if err := r.createKubeconfig(ctx, kubeconfigSecret, clusterName, acp); err != nil {
-					log.Error(err, "failed to create secret for agent control plane", "agent cluster install name", aci.Name, "agent cluster install namespace", aci.Namespace, "agent control plane name", acp.Name)
-					return ctrl.Result{}, err
-				}
-			}
-			// Update AgentControlPlane status
-			acp.Status.Ready = true
-			if err := r.Client.Status().Update(ctx, &acp); err != nil {
-				log.Error(err, "failed updating agent control plane to ready for agentclusterinstall", "agent cluster install name", aci.Name, "agent cluster install namespace", aci.Namespace, "agent control plane name", acp.Name)
-				return ctrl.Result{}, err
-			}
+	if !r.ClusterKubeconfigSecretExists(ctx, clusterName, acp.Namespace) {
+		if err := r.createKubeconfig(ctx, kubeconfigSecret, clusterName, *acp); err != nil {
+			return err
 		}
 	}
 
-	return ctrl.Result{}, nil
+	acp.Status.Initialized = true
+	if err := r.Client.Status().Update(ctx, acp); err != nil {
```

Comment:
WDYT about moving to patcher update on deferred method as other controllers are doing?

Reply:
Nevermind it doesnt make sense in this case